### PR TITLE
Only call hooks when setState has finished updating the internal state

### PR DESCRIPTION
**Pull request by DaanDD**
### Description
Before this, when a hook like `onViewModeChange`, `onNavigateBack` or `onNavigateForward` was called, the internal state would not yet be updated. This meant that when I tried to access the current `viewDate` through the internal state after one of these hooks was called, the setState would not have been updated yet, because `setState` is asynchronous. This PR only calls these hooks when `setState` has completed.

### Checklist
```
[x] I have not included any built dist files (us maintainers do that prior to a new release)
[x] I have added tests covering my changes
[x] All new and existing tests pass
[ ] My changes required the documentation to be updated
  [ ] I have updated the documentation accordingly
  [ ] I have updated the TypeScript 1.8 type definitions accordingly
  [ ] I have updated the TypeScript 2.0+ type definitions accordingly
```

The tests had to be refactored a bit. I had to implement an async/await pattern that waits for the `setState` to complete before moving on to the next test. For some reason all tests within the hooks are ignored without the async/await pattern. This also meant I had to upgrade the eslint and the eslint parser for the tests to ES2017. This also meant I had to fix some issues in the code itself.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -51,10 +51,11 @@
     "babel-loader": "^6.2.1",
     "babel-plugin-transform-remove-strict-mode": "0.0.2",
     "babel-preset-es2015": "^6.22.0",
+    "babel-preset-2017": "0.0.0",
     "babel-preset-react": "^6.22.0",
     "enzyme": "^3.0.0",
     "enzyme-adapter-react-15": "^1.0.5",
-    "eslint": "^3.1.0",
+    "eslint": "^4.17.0",
```
Comment (simeg) [inline package.json]: There seems to be a lot of changes completely unrelated to the commit message in this commit. It would be nice to have them separated, like why is this dependency updated? And why is `babel-preset-2017` introduced?

Comment (DaanDD) [reply package.json]: It seems like I committed a bit too much. I thought the es2017 preset for `eslint` could only work in newer versions, but turns out it didn't. The `babel-preset-2017` also didn't belong there. I have cleaned it up accordingly.

File: test/.eslintrc.js [hunk 2]
```diff
@@ -1,6 +1,6 @@
 module.exports = {
 		"parserOptions": {
-				"ecmaVersion": 6,
+				"ecmaVersion": 2017,
```
Comment (simeg) [inline test/.eslintrc.js]: Why is this needed?

Comment (DaanDD) [reply test/.eslintrc.js]: Because I needed to use async/await in the tests to wait for the `setState` callback to complete.

**Comment by arqex** [comment]
Sorry for getting late into this PR but thanks!. Many changes happened since it was created and the issues are already fixed in the latest version. Closing it.
